python/triton/language/semantic.py

```diff
@@ -1677,6 +1677,31 @@ def associative_scan(inputs: Sequence[tl.tensor], axis: int, region_builder_fn,
     return tuple(wrap_tensor(scan_op.get_result(i), inputs[i].type.scalar, shape) for i in range(len(inputs)))
 
 
+# ===----------------------------------------------------------------------===
+#                               Gather
+# ===----------------------------------------------------------------------===
+
+
+def gather(src: tl.tensor, index: tl.tensor, axis: int, builder: ir.builder) -> tl.tensor:
+    assert index.dtype.is_int(), "index must be an integer tensor"
+
+    rank = len(src.type.shape)
+    assert len(index.type.shape) == rank, "source and index tensors must have the same rank"
+
+    assert -rank <= axis < rank, f"gather axis {axis} must be < source rank ({rank})"
+    if axis < 0:
+        axis += rank
+
+    for d in range(rank):
+        if d == axis:
+            continue
+        assert index.type.shape[d] <= src.type.shape[
+            d], f"index dim {axis} cannot be greater than the corresponding source dim"
```

Comment:
This is a bit strange. You're allowing the gather op to implicitly slice the `src` tensor to match the `index` tensor? If we're going to allow this I think it should be its own operation.

Reply:
I thought that's what we wanted? 

I guess broadcasting could be supported

Reply:
> I thought that's what we wanted?

I 100% expect that the gather axis can be a different shape, but's it's not normal for the other dimensions to be allowed to be a different shape. I find it very surprising coming from numpy/pytorch semantics.

Also I don't think this behavior is compatible with broadcasting as it would be ambiguous. If the index dimension is  of size 1 we can't tell if it's supposed to be a slice, or if it should be broadcasted.

Reply:
I suppose there is an advantage to fusing the gather op with the slice as in general I think I slice op could have to go through shared memory to transfer redundant data. Perhaps this could be a pattern matched lowering instead of implicit behavior of `tt.gather` though?

Reply:
Huh I guess [`torch.gather`](https://pytorch.org/docs/stable/generated/torch.gather.html) actually does have this behavior. Now I'm not sure what to think haha.  It feels wrong that there are huge chunks of the input tensor that get completely ignored, and feels to me like two operations. 

I'm also a bit confused what the use case for this would be, as there's no way to create a slice that doesn't start at 0.

Reply:
I don't really have an opinion here on the semantics of `tl.gather`, so let me know what you two prefer!

Reply:
Am I understanding correctly that the output, source, and index tensor should all have the same shape?

Reply:
I'm voting to keep it simple and have all the dims other than the specified `dim` to be equal to the tensor dims. We can always relax this in the future if there's a use case (I don't know of one ATM).

Reply:
>Am I understanding correctly that the output, source, and index tensor should all have the same shape?

Not quite. They should their shape should match in all the dimensions but the one that we are gathering. `gather` is just the "batched version" of doing the following 1D advanced indexing
```
source = torch.randn((N,))
idx = torch.randint(N, (K,))
output = source[idx]
```

Reply:
Okay thanks for chiming in! Let me alter the restriction.

Reply:
I altered the builtin and op's invariants to require that the index and output dims other than the gather dim match the source dims, ptal